File: android/crosspl/anno/src/main/kotlin/org/elastos/tools/crosspl/processor/CrossVariableType.kt
package org.elastos.tools.crosspl.processor

import javax.lang.model.element.TypeElement
import javax.lang.model.type.ArrayType
import javax.lang.model.type.DeclaredType
import javax.lang.model.type.TypeKind
import javax.lang.model.type.TypeMirror

enum class CrossVariableType {
    BOOLEAN,
    INTEGER,
    LONG,
    DOUBLE,
    VOID,

    STRING,
    BYTEARRAY,
    FUNCTION,

    STRINGBUFFER,
    BYTEBUFFER,

    CROSSBASE;

    companion object {
        fun Parse(type: TypeMirror): CrossVariableType? {
            if(supportedTypeKind.contains(type.kind)) {
                return supportedTypeKind.getValue(type.kind)
            } else if(type is DeclaredType) {
                val typeElement = type.asElement() as TypeElement
                val typeStr = typeElement.toString()
                return supportedTypeDeclared[typeStr]
            } else if(type is ArrayType) {
                val typeStr = type.toString()
                return supportedTypeDeclared[typeStr]
            }

            return null
        }

        private val supportedTypeKind = mapOf(
            TypeKind.BOOLEAN           to BOOLEAN,
            TypeKind.INT               to INTEGER,
            TypeKind.LONG              to LONG,
            TypeKind.DOUBLE            to DOUBLE,
            TypeKind.VOID              to VOID
        )
        private val supportedTypeDeclared = mapOf(
            "java.lang.String"                    to STRING,
            "byte[]"                              to BYTEARRAY,
            "java.lang.Runnable"                  to FUNCTION,
            "java.lang.StringBuffer"              to STRINGBUFFER,
            "java.io.ByteArrayOutputStream"       to BYTEBUFFER,
            "org.elastos.tools.crosspl.CrossBase" to CROSSBASE
        )
    }

    fun isPrimitiveType(): Boolean {
        val primitiveTypeSet = setOf(
            BOOLEAN,
            INTEGER,
            LONG,
            DOUBLE,
            VOID
        )

        return primitiveTypeSet.contains(this)
    }

    fun toCppString(isConst: Boolean = false): String {
        val primitiveTypeMap = mapOf(
            BOOLEAN    to "bool",
            INTEGER    to "int",
            LONG       to "int64_t",
            DOUBLE     to "double",
            VOID       to "void"
        )
        val classTypeMap = mapOf(
            STRING       to "const char*",
            BYTEARRAY    to "std::span<int8_t>",
            FUNCTION     to "std::function<void()>",
            STRINGBUFFER to "std::stringstream",
            BYTEBUFFER   to "std::vector<int8_t>",
            CROSSBASE    to "::CrossBase"
        )

        var cppType = toString(primitiveTypeMap, classTypeMap, isConst)

        return cppType
    }

    fun toJniString(isConst: Boolean = false): String {
        val primitiveTypeMap = mapOf(
            BOOLEAN    to "jboolean",
            INTEGER    to "jint",
            LONG       to "jlong",
            DOUBLE     to "jdouble",
            VOID       to "void"
        )
        val classTypeMap = mapOf(
            STRING       to "jstring",
            BYTEARRAY    to "jbyteArray",
            FUNCTION     to "jobject",
            STRINGBUFFER to "jobject",
            BYTEBUFFER   to "jobject",
            CROSSBASE    to "jobject"
        )

        var jniType = toString(primitiveTypeMap, classTypeMap, isConst)

        return jniType
    }

    fun toJniSigChar(): String {
        val primitiveTypeMap = mapOf(
            BOOLEAN    to "Z",
            INTEGER    to "I",
            LONG       to "J",
            DOUBLE     to "D",
            VOID       to "V"
        )
        val classTypeMap = mapOf(
            STRING       to "Ljava/lang/String;",
            BYTEARRAY    to "[B",
            FUNCTION     to "Ljava/lang/Runnable;",
            STRINGBUFFER to "Ljava/lang/StringBuffer;",
            BYTEBUFFER   to "Ljava/io/ByteArrayOutputStream;",
            CROSSBASE    to "Lorg/elastos/tools/crosspl/CrossBase;"
        )

        var javaChar = toString(primitiveTypeMap, classTypeMap)

        return javaChar
    }

    fun toString(primitiveTypeMap: Map<CrossVariableType, String>,
                 classTypeMap: Map<CrossVariableType, String>,
                 isConst: Boolean = false): String {
        var cppType = primitiveTypeMap[this]
        if(cppType == null) {
            cppType = classTypeMap[this]
        }

        if(isConst
        && ! primitiveTypeMap.contains(this)
        && this != STRING) {
            cppType = "const ${cppType!!}*"
        }

        return cppType!!
    }

    override
    fun toString(): String {
        val stringMap = mapOf(
            BOOLEAN      to "Boolean",
            INTEGER      to "Int",
            LONG         to "Long",
            DOUBLE       to "Double",
            VOID         to "Void",
            STRING       to "String",
            BYTEARRAY    to "ByteArray",
            FUNCTION     to "Function",
            STRINGBUFFER to "StringBuffer",
            BYTEBUFFER   to "ByteBuffer",
            CROSSBASE    to "CrossBase"
        )

        return stringMap.getValue(this)
    }
}

File: ios/anno/swift/CrossVariableType.swift
import Foundation

class CrossVariableType {
  public enum `Type` {
    case BOOLEAN
    case INT32
    case INT64
    case DOUBLE
    case VOID
    
    case STRING
    case BYTEARRAY
    case FUNCTION
    
    case STRINGBUFFER
    case BYTEBUFFER
    
    case CROSSBASE
  }
  
  static func Parse(sourceContent: String) -> CrossVariableType {
    let supportedTypeKind = ["Bool": Type.BOOLEAN,
                             "Int32": Type.INT32,
                             "Int64": Type.INT64,
                             "Double": Type.DOUBLE,
                             "Void": Type.VOID]
    
    let supportedTypeDeclared = ["String": Type.STRING,
                                 "Data": Type.BYTEARRAY,
                                 "Function": Type.FUNCTION,
                                 
                                 "inout String": Type.STRINGBUFFER,
                                 "inout NSString": Type.STRINGBUFFER,
                                 "inout Data": Type.BYTEBUFFER,
                                 "inout NSData": Type.BYTEBUFFER,
                                 
                                 "CrossBase": Type.CROSSBASE]
    
    var trimContent = sourceContent.trimmingCharacters(in: .whitespacesAndNewlines)
   
    let varType = CrossVariableType()
    varType.type = supportedTypeKind[trimContent]
    if varType.type != nil {
      return varType
    }
    
    trimContent = trimContent.trimmingCharacters(in: CharacterSet(charactersIn: "?"))
    varType.type = supportedTypeDeclared[trimContent]
    if varType.type == nil {
      print("CrossVariableType.Parse() Unsupported variable type: \(trimContent)")
      exit(1)
    }
    
//    print("=================CrossVariableType.Parse() trimContent=\(varType.toString())")

    return varType
  }
  
  func toString() -> String {
    let typeMap = [
      Type.BOOLEAN: "Bool",
      Type.INT32: "Int32",
      Type.INT64: "Int64",
      Type.DOUBLE: "Double",
      Type.VOID: "Void",
      
      Type.STRING: "String",
      Type.BYTEARRAY: "ByteArray",
      Type.FUNCTION: "Function",
      Type.STRINGBUFFER: "StringBuffer",
      Type.BYTEBUFFER: "ByteBuffer",
      Type.CROSSBASE: "CrossBase"
    ]
    
    //  var cppType = toString(primitiveTypeMap, classTypeMap, isConst)
    let strType = typeMap[type!]
    
    return strType!
  }
  
  func isPrimitiveType() -> Bool {
    let primitiveTypeSet = [
      Type.BOOLEAN,
      Type.INT32,
      Type.INT64,
      Type.DOUBLE,
      Type.VOID
    ]
  
    return primitiveTypeSet.contains(type!)
  }

  
  func toCppString(isConst: Bool = false) -> String {
    let typeMap = [
      Type.BOOLEAN: "bool",
      Type.INT32: "int32_t",
      Type.INT64: "int64_t",
      Type.DOUBLE: "double",
      Type.VOID: "void",

      Type.STRING: "const char*",
      Type.BYTEARRAY: "std::span<int8_t>",
      Type.FUNCTION: "std::function<void()>",
      Type.STRINGBUFFER: "std::stringstream",
      Type.BYTEBUFFER: "std::vector<int8_t>",
      Type.CROSSBASE: "::CrossBase"
    ]
  
//  var cppType = toString(primitiveTypeMap, classTypeMap, isConst)
    let cppType = typeMap[type!]
    
    return cppType!
  }
  
  func toObjcString(isConst: Bool = false) -> String {
    let typeMap = [
      Type.BOOLEAN: "bool",
      Type.INT32: "int32_t",
      Type.INT64: "int64_t",
      Type.DOUBLE: "double",
      Type.VOID: "void",

      Type.STRING: "NSString*",
      Type.BYTEARRAY: "NSData*",
      Type.FUNCTION: "---",
      Type.STRINGBUFFER: "NSString**",
      Type.BYTEBUFFER: "NSData**",
      Type.CROSSBASE: "NSObject*"
    ]
    
//    var jniType = toString(primitiveTypeMap, classTypeMap, isConst)
    let objcType = typeMap[type!]
    
    return objcType!
  }

  func toSwiftString() -> String {
    let typeMap = [
      Type.BOOLEAN: "Bool",
      Type.INT32: "Int32",
      Type.INT64: "Int64",
      Type.DOUBLE: "Double",
      Type.VOID: "Void",
      
      Type.STRING: "String",
      Type.BYTEARRAY: "Data",
      Type.FUNCTION: "---",
      Type.STRINGBUFFER: "inout String",
      Type.BYTEBUFFER: "inout Data",
      Type.CROSSBASE: "CrossBase"
    ]
  
    let swiftType = typeMap[type!]
    
    return swiftType!
  }

  
  var type: Type?
}
